File: 209.rb
def min_sub_array_len(s, nums)
  return 0 if nums.length == 0

  sum = nums[0]
  liter, riter = 0, 0
  min_size = nums.length + 1

  (riter += 1; sum += nums[riter]) while sum < s && riter < nums.length - 1

  return 0 if sum < s && riter == nums.length - 1
  min_size = [min_size, riter - liter + 1].min

  while true
    (sum -= nums[liter]; liter += 1) while liter < riter && sum - nums[liter] >= s
    min_size = [min_size,  riter - liter + 1].min
    break if riter == nums.length - 1
    (riter += 1; sum += nums[riter])
  end

  min_size
end

File: 209.cpp
#include <limits>
#include <vector>


int minSubArrayLenSlow(int s, const std::vector<int>& nums) {
    const int LEN = nums.size();
    if (LEN == 0) return 0;

    int i = 0;
    int sum = 0;
    int minLen = std::numeric_limits<int>::max();

    for(int j = 1; j <= LEN; ++j) {
        sum += nums[j - 1];
        while(sum >= s && i <= j - 1) {
            minLen = std::min(minLen, j - i);
            sum -= nums[i++];
        }
    }

    if (minLen == std::numeric_limits<int>::max()) {
        return 0;
    }
    return minLen;

}


int minSubArrayLen(int s, const std::vector<int>& nums) {
    const int LEN = nums.size();
    if (LEN == 0) return 0;

    int i = 0;
    int minLen = std::numeric_limits<int>::max();

    for(int j = 1; j <= LEN; ++j) {
        s -= nums[j - 1];
        while(s <= 0 && i <= j - 1) {
            minLen = std::min(minLen, j - i);
            s += nums[i++];
        }
    }

    if (minLen == std::numeric_limits<int>::max()) {
        return 0;
    }
    return minLen;
}
